### PR TITLE
More accurate delay() function from BenF.

### PR DESCRIPTION
**Pull request by Rocketct**
porting of https://github.com/arduino/ArduinoCore-avr/commit/67c0a1995aace2953e481ef6183b1b91e04942fc#diff-ab397febf179e1982f4ee0758df4c982

cc/ @sandeepmistry tested it works

### CODE REVIEW

**Comment by sandeepmistry** [comment]
Resolves: https://github.com/arduino/ArduinoCore-samd/issues/356

**Comment by sandeepmistry** [comment]
@Rocketct the code change looks fine, what sketch(es) did you use to test this?

**Comment by Rocketct** [comment]
i have used the one in the forum with was open the original issue on the AVR : forum.arduino.cc/index.php?topic=22902.15 and compared the results obtained by using a MKRZERO and a UNO, the sketch is the following:
```c++
/* Test delay and delayMicroseconds */

#define NUM_TESTS 100
#define MAX_DELAY 20

void delay_debugged(unsigned long ms)
{
     unsigned long start = millis();
     
  /* Original from wiring.c:  
     while (millis() - start <= ms)
  */
     while (millis() - start < ms)
           ;
}

void setup()   {                
 Serial.begin(115200);
}

void loop()                    
{
  int i, j, k;
  unsigned long tic, toc;
  float avg_delay;
   
  for (i=1; i<=MAX_DELAY; i++)
  {
     avg_delay = 0;
     for (j=0; j<NUM_TESTS; j++)
     {
        tic = micros();
        delay(i);
        toc = micros();
        avg_delay += (toc - tic);
     }
     avg_delay /= (1000.0 * NUM_TESTS);
     Serial.print("delay(");Serial.print(i);Serial.print(")=");
     Serial.println(avg_delay);
  }

  for (i=1; i<=MAX_DELAY; i++)
  {
     avg_delay = 0;
     for (j=0; j<NUM_TESTS; j++)
     {
        tic = micros();
        delay_debugged(i);
        toc = micros();
        avg_delay += (toc - tic);
     }
     avg_delay /= (1000.0 * NUM_TESTS);
     Serial.print("delay_debugged(");Serial.print(i);Serial.print(")=");
     Serial.println(avg_delay);
  }

  for (i=1; i<=MAX_DELAY; i++)
  {
     avg_delay = 0;
     for (j=0; j<NUM_TESTS; j++)
     {
        k = 4*i;
        tic = micros();
        delayMicroseconds(k);
        toc = micros();
        avg_delay += (toc - tic);
     }
     avg_delay /= NUM_TESTS;
     Serial.print("delayMicrosecnds(");Serial.print(k);Serial.print(")=");
     Serial.println(avg_delay);
  }

  Serial.println("-----------------------------------------------");
  delay(5000);
}```

File: cores/arduino/delay.c [hunk 1]
```diff
@@ -68,12 +68,15 @@ void delay( unsigned long ms )
...
-
-  do
-  {
-    yield() ;
-  } while ( _ulTickCount - start < ms ) ;
+  uint32_t start = micros();
+
+   while (ms > 0) {
+     yield();
+     while ( ms > 0 && (micros() - start) >= 1000) {
```
Comment (cmaglie) [inline cores/arduino/delay.c]: - indent is off by 1 space for all the function except the first line
- there is an extra space after the open parenthesis before `ms`

Comment (Rocketct) [reply cores/arduino/delay.c]: @cmaglie removed the white space!

**Comment by sslupsky** [comment]
I there a problem with this PR? Every time micros() rolls over, you will loose a millisecond tick because (micros() - start) has a negative result? This happens once every approximately 4.29 seconds.

This comment is cross posted to a comparison between 1.6.19 and 1.6.21

**Pull request merged**